components/observations/uploader/ObservationsUploader.ts

```diff
@@ -156,6 +265,65 @@ export class ObservationUploader {
     this.tryRunTaskQueue();
   }
 
+  private findTaskById(taskId: string) {
+    return this.completedTasks.find(t => t.id === taskId) || this.taskQueue.find(t => t.id === taskId);
+  }
+
+  private findTasksByParentId(taskId: string) {
+    return [...this.completedTasks.filter(t => t.parentId === taskId), ...this.taskQueue.filter(t => t.parentId === taskId)];
+  }
+
+  private setTaskStatus(taskId: string, status: TaskStatus, updateChildren = true) {
+    const task = this.findTaskById(taskId);
+
+    if (!task) {
+      this.logger.warn({taskId, status}, 'Unable to update task status - task not found');
+      return false;
+    } else if (task.status === 'working') {
+      this.logger.warn({taskId, status}, 'Unable to update task status - task is in progress');
+      return false;
+    } else if (task.status === 'error' || task.status === 'success') {
+      this.logger.warn({taskId, status}, 'Unable to update task status - task has finished');
+      return false;
+    }
+
+    task.status = status;
+
+    if (updateChildren) {
+      this.findTasksByParentId(taskId).map(t => this.setTaskStatus(t.id, status, updateChildren));
+    }
+
+    return true;
+  }
+
+  pauseUpload(observationId: string, pauseImages = true) {
+    this.setTaskStatus(observationId, 'paused', pauseImages);
+    this.tryRunTaskQueue();
+  }
+
+  resumeUpload(observationId: string, resumeImages = true) {
+    this.setTaskStatus(observationId, 'pending', resumeImages);
+    this.tryRunTaskQueue();
+  }
+
+  getUploadStatus(observationId: string): UploadStatus {
+    const task = this.findTaskById(observationId);
+    if (!task) {
+      throw new Error(`No task found with id ${observationId}`);
+    }
+    return {
+      networkStatus: this.offline ? 'offline' : 'online',
+      observation: {
+        status: task.status,
+        attemptCount: task.attemptCount,
+      },
+      images: this.findTasksByParentId(observationId).map(t => ({
+        status: t.status,
```

Comment:
What's the implication of doing a copy for the values here? Do we need to worry that the caller gets a snapshot of the status at call-time? I don't fully grok the runtime here but will the caller ever process stale data after `getUploadStatus` (can the routine be preempted?)

Reply:
1. I am not sure I understand the root of this question. I think you're looking for something more in an answer beyond "it's a copy", but you will have to help me.
2. It's explicitly designed to be a snapshot, so...I guess I'm not worried. 
3. I'm afraid I don't understand this question, sorry - what "stale data", which "routine"? in general no, there is no preemption - but async code flows can mean that data is modified across `await` boundaries.

maybe you can give me an example of a problem you're worried we might hit?

Reply:
I guess I was considering the case where:

1. caller uses `getUploadStatus()`, gets a snapshot/pass-by-value copy
2. preemption occurs, other code runs and one or more tasks in the status change status
3. caller returns to execution and operates over stale data - perhaps using stale status to re-try something etc

Does that matter?

Reply:
(Or is my mental model of the runtime here just wrong, and this is not important?)

Reply:
the JS runtime doesn't allow preemption in the traditional sense so it's less of a concern. if your code is running, then it runs until you hit the end of your scope, as long as you don't do anything `async`.

that is to say, this is safe

```js
const status = getUploader().getUploadStatus(id);
doSomethingWithoutAwaiting();
// status can't be stale here
```

but this is less safe
```js
const status = getUploader().getUploadStatus(id);
await doSomethingAsync();
// status might be stale at this point, because you've yielded 
// and `processTaskQueue` will get a chance to run if it's waiting
```

I think the class of bugs you're describing are still possible, they're just less likely than in trad multithreaded environments.

Reply:
Makes sense, so it's on the caller to be careful then and this function does not have to be concerned.

Reply:
Thanks for the explanation!